# Project Work Part 1

#### a)
Downloading data and displaying some of it

In [156]:
import numpy as np
import polars as pl

# read data
links_df = pl.read_csv("../data/links.csv", schema={"movieId": pl.Int32, "imdb": pl.Int32, "tmdbId": pl.Int32})
movies_df = (
    pl.read_csv("../data/movies.csv", schema={"movieId": pl.Int32, "title": pl.String, "genres": pl.String})
    .with_row_index()
    .with_columns(
        pl.col("index")
        .cast(pl.Int32)
    )
)
ratings_df = pl.read_csv("../data/ratings.csv", schema={"userId": pl.Int32, "movieId": pl.Int32, "rating": pl.Float32, "timestamp": pl.Int32})
tags_df = pl.read_csv("../data/tags.csv", schema={"userId": pl.Int32, "movieId": pl.Int32, "tag": pl.String, "timestamp": pl.Int32})

print("Number of ratings: ", ratings_df.height)

Number of ratings:  100836


In [157]:
print(links_df)

shape: (9_742, 3)
┌─────────┬─────────┬────────┐
│ movieId ┆ imdb    ┆ tmdbId │
│ ---     ┆ ---     ┆ ---    │
│ i32     ┆ i32     ┆ i32    │
╞═════════╪═════════╪════════╡
│ 1       ┆ 114709  ┆ 862    │
│ 2       ┆ 113497  ┆ 8844   │
│ 3       ┆ 113228  ┆ 15602  │
│ 4       ┆ 114885  ┆ 31357  │
│ 5       ┆ 113041  ┆ 11862  │
│ …       ┆ …       ┆ …      │
│ 193581  ┆ 5476944 ┆ 432131 │
│ 193583  ┆ 5914996 ┆ 445030 │
│ 193585  ┆ 6397426 ┆ 479308 │
│ 193587  ┆ 8391976 ┆ 483455 │
│ 193609  ┆ 101726  ┆ 37891  │
└─────────┴─────────┴────────┘


In [158]:
print(movies_df)

shape: (9_742, 4)
┌───────┬─────────┬─────────────────────────────────┬─────────────────────────────────┐
│ index ┆ movieId ┆ title                           ┆ genres                          │
│ ---   ┆ ---     ┆ ---                             ┆ ---                             │
│ i32   ┆ i32     ┆ str                             ┆ str                             │
╞═══════╪═════════╪═════════════════════════════════╪═════════════════════════════════╡
│ 0     ┆ 1       ┆ Toy Story (1995)                ┆ Adventure|Animation|Children|C… │
│ 1     ┆ 2       ┆ Jumanji (1995)                  ┆ Adventure|Children|Fantasy      │
│ 2     ┆ 3       ┆ Grumpier Old Men (1995)         ┆ Comedy|Romance                  │
│ 3     ┆ 4       ┆ Waiting to Exhale (1995)        ┆ Comedy|Drama|Romance            │
│ 4     ┆ 5       ┆ Father of the Bride Part II (1… ┆ Comedy                          │
│ …     ┆ …       ┆ …                               ┆ …                               │
│ 9737  ┆ 1935

In [159]:
print(ratings_df)

shape: (100_836, 4)
┌────────┬─────────┬────────┬────────────┐
│ userId ┆ movieId ┆ rating ┆ timestamp  │
│ ---    ┆ ---     ┆ ---    ┆ ---        │
│ i32    ┆ i32     ┆ f32    ┆ i32        │
╞════════╪═════════╪════════╪════════════╡
│ 1      ┆ 1       ┆ 4.0    ┆ 964982703  │
│ 1      ┆ 3       ┆ 4.0    ┆ 964981247  │
│ 1      ┆ 6       ┆ 4.0    ┆ 964982224  │
│ 1      ┆ 47      ┆ 5.0    ┆ 964983815  │
│ 1      ┆ 50      ┆ 5.0    ┆ 964982931  │
│ …      ┆ …       ┆ …      ┆ …          │
│ 610    ┆ 166534  ┆ 4.0    ┆ 1493848402 │
│ 610    ┆ 168248  ┆ 5.0    ┆ 1493850091 │
│ 610    ┆ 168250  ┆ 5.0    ┆ 1494273047 │
│ 610    ┆ 168252  ┆ 5.0    ┆ 1493846352 │
│ 610    ┆ 170875  ┆ 3.0    ┆ 1493846415 │
└────────┴─────────┴────────┴────────────┘


In [160]:
print(tags_df)

shape: (3_683, 4)
┌────────┬─────────┬──────────────────┬────────────┐
│ userId ┆ movieId ┆ tag              ┆ timestamp  │
│ ---    ┆ ---     ┆ ---              ┆ ---        │
│ i32    ┆ i32     ┆ str              ┆ i32        │
╞════════╪═════════╪══════════════════╪════════════╡
│ 2      ┆ 60756   ┆ funny            ┆ 1445714994 │
│ 2      ┆ 60756   ┆ Highly quotable  ┆ 1445714996 │
│ 2      ┆ 60756   ┆ will ferrell     ┆ 1445714992 │
│ 2      ┆ 89774   ┆ Boxing story     ┆ 1445715207 │
│ 2      ┆ 89774   ┆ MMA              ┆ 1445715200 │
│ …      ┆ …       ┆ …                ┆ …          │
│ 606    ┆ 7382    ┆ for katie        ┆ 1171234019 │
│ 606    ┆ 7936    ┆ austere          ┆ 1173392334 │
│ 610    ┆ 3265    ┆ gun fu           ┆ 1493843984 │
│ 610    ┆ 3265    ┆ heroic bloodshed ┆ 1493843978 │
│ 610    ┆ 168248  ┆ Heroic Bloodshed ┆ 1493844270 │
└────────┴─────────┴──────────────────┴────────────┘


#### b)
User-based collaborative filtering using the Pearson correlation function.

The pearson correlation function is found in polars so there is no need to implement it from scratch. The funciton will take into account only the ratings that are rated by both users. This is done by filtering out the ratings that are not rated by both users ignoring nan values.

Let's calculate top similar users for userId 1 using Pearson correlation


In [161]:
def get_top_n_most_similar_users(user_id, n):
    correlation_df = (
        ratings_df
        .drop("timestamp")
        .pivot("userId", index="movieId")
        .drop("movieId")
        .select(pl.corr(pl.all(), str(user_id), method="pearson"))
        .transpose(include_header=True, header_name="userId", column_names=["correlation"])
        .cast({"userId": pl.Int16})
        .filter(pl.col("correlation").is_not_nan()) # filter out NaN values
        .sort(["correlation", "userId"], descending=[True, False])
        .filter(pl.col("userId") != user_id) # filter out the user itself
        .filter(pl.col("correlation") > 0.2) # filter out users with low correlation
        .limit(n)
    )
    return (
        correlation_df
        .select("userId")
        .to_numpy()
        .reshape(-1)
    ), (
        correlation_df
        .select("correlation")
        .to_numpy()
        .reshape(-1)
    )

# Calculate the top 50 most similar users to userId 1
user_ids, similarity_scores = get_top_n_most_similar_users(1, 50)
print(user_ids)
print(similarity_scores)

[106 146 333 550 598 473 511   9  13 366 401 535 154 481  90 499 157 139
 476 487 210 114 530  49 162 297 207  44 430 394 574 421 248 173  65 505
 369 435 375 431 351  72 206 120 344 112  99 445 494 467]
[1.         1.         1.         1.         1.         0.9622505
 0.9258201  0.91855866 0.8783101  0.8728716  0.8669214  0.86640024
 0.8660254  0.8660254  0.8215838  0.8029551  0.80178374 0.7905694
 0.78693587 0.7745967  0.7676495  0.7592566  0.75592893 0.75
 0.7083333  0.7062815  0.7016464  0.6844476  0.65465367 0.6506001
 0.6437963  0.64122343 0.62469506 0.6172134  0.61237246 0.61237246
 0.61209774 0.6109598  0.6061281  0.6030227  0.6        0.5964321
 0.5962848  0.59487444 0.58760905 0.58389246 0.58345    0.56906945
 0.5673498  0.5669467 ]


Pearson correlation function has two major problems. If the standard deviation of either of the ratings is 0 the function will give nan result as there is a division by zero. This is a problem since in a case where the user has given same score for all movies the standard deviation will be 0 and the function will not work. For example users with ratings [5, 5, 5, 5] and [5, 5, 5, 5] will have a correlation of nan even though they are exactly similar. Also users with low and high ratings e.g. [5, 4, 5, 5] and [2, 1, 2, 2] will still have high Pearson correlation even though they are not similar at all.

In [162]:
# Example of the problems with Pearson correlation
df = pl.DataFrame(
    {
    "a": [5, 5, 5, 5],
    "b": [5, 5, 5, 5],
    "c": [5, 4, 5, 5],
    "d": [2, 1, 2, 2],
    }
)

print(df.select(pl.corr("a", "b", method="pearson")))
print(df.select(pl.corr("c", "d", method="pearson")))

shape: (1, 1)
┌─────┐
│ a   │
│ --- │
│ f64 │
╞═════╡
│ NaN │
└─────┘
shape: (1, 1)
┌─────┐
│ c   │
│ --- │
│ f64 │
╞═════╡
│ 1.0 │
└─────┘


#### c)
Prediction function from course slides

In [163]:
def predict_movies(user_ratings_mean, similar_users_ratings, similar_users_ratings_mean, similarity_scores, candidate_movie_indices, num_of_movies):
    """
    The prediction function from course slides.

    :param user_ratings_mean: mean of user ratings
    :param similar_users_ratings: sparse numpy 2D matrix of similar users ratings. Each row is a user and each column is a movie
    :param similar_users_ratings_mean: numpy array of mean ratings of the similar users
    :param similarity_scores: numpy array of similarity scores between the user and the similar users
    :param candidate_movie_indices: numpy array of movie indices that the similar users have rated
    :param num_of_movies: number of movies to recommend
    :return: top ranked candidate movie indices, and their predicted ratings
    """
    
    # Compute ratings only for the candidate movies to save computation
    candidate_similar_users_ratings = np.take(similar_users_ratings, candidate_movie_indices, axis=1)

    # Subtract mean rating from each rating, (r_b,p - r_b)
    delta_ratings = candidate_similar_users_ratings - similar_users_ratings_mean.reshape(similar_users_ratings_mean.shape[0], 1)

    # Sum sim(a,b) * (r_b,p - r_b)
    numerator = np.sum(similarity_scores.reshape(-1, 1) * delta_ratings, axis=0)

    # Sum |sim(a,b)|, taking absolute value to avoid negative bias even though it should not be possible
    denominator = np.sum(np.abs(similarity_scores))

    # Sum sim(a,b) * (r_b,p - r_b) / Sum |sim(a,b)|
    bias = numerator / denominator

    # pred(a,p) = r_a + bias
    predicted_ratings = user_ratings_mean + bias

    # Rank the movies by predicted rating
    indices = np.flip(np.argsort(predicted_ratings))
    predicted_movie_indices = candidate_movie_indices[indices][:num_of_movies]
    sorted_ratings = predicted_ratings[indices][:num_of_movies]

    return predicted_movie_indices, sorted_ratings

Let's predict movies for userId 1 using the top 50 most similar users and the prediction function from the course slides

In [176]:
# Define some helper functions to extract data from the dataframes

# ratings
def get_ratings_for_users(user_ids):
    return (
        ratings_df
        .drop("timestamp")
        .sort("movieId")
        .pivot("userId", index="movieId")
        .drop("movieId")
        .select([str(user_id) for user_id in user_ids])
        .transpose(include_header=True, header_name="userId", column_names=user_ids)
        .drop("userId")
        .to_numpy()
    )
user_ids = np.array([1, 2, 3, 4, 5])
print(get_ratings_for_users(user_ids))

# ratings mean
def get_mean_ratings_for_users(user_ids):
    return (
        ratings_df
        .drop("timestamp")
        .sort("movieId")
        .pivot("userId", index="movieId")
        .drop("movieId")
        .select([str(user_id) for user_id in user_ids.tolist()])
        .select(pl.all().mean())
        .transpose(include_header=True, header_name="userId", column_names=["ratings_mean"])
        .cast({"userId": pl.Int16})
        .filter(pl.col("userId").is_in(user_ids))
        .select("ratings_mean")
        .to_numpy()
        .reshape(-1)
    )
print(get_mean_ratings_for_users(user_ids))

# movie indices for movie ids
def get_movie_indices_for_movie_ids(movie_ids):
    return (
        movies_df
        .filter(pl.col("movieId").is_in(movie_ids))
        .select("index")
        .to_numpy()
        .reshape(-1)
    )
print(get_movie_indices_for_movie_ids(np.array([1, 2, 3, 100, 112])))

# single user rating mean
def get_user_rating_mean(user_id):
    return get_mean_ratings_for_users(np.array([user_id]))[0]

# candidate movie ids
def get_candidate_movie_ids(user_id, similar_user_ids):
    rated_by_user = (
        ratings_df
        .filter(pl.col("userId") == user_id)
        .select("movieId")
        .to_numpy()
        .reshape(-1)
    )

    return (
        ratings_df
        .filter(
            pl.col("userId").is_in(similar_user_ids), # only include ratings from similar users
            pl.col("movieId").is_in(rated_by_user).not_() # only include movies that the user has not rated
        )
        .group_by("movieId")
        .agg()
        .select("movieId")
        .to_numpy()
        .reshape(-1)
    )

print(get_candidate_movie_ids(1, np.array([2, 3, 4, 5, 6]))[:10])

[[ 4. nan  4. ... nan nan nan]
 [nan nan nan ... nan nan nan]
 [nan nan nan ... nan nan nan]
 [nan nan nan ... nan nan nan]
 [ 4. nan nan ... nan nan nan]]
[4.3663793 3.9482758 2.4358974 3.5555556 3.6363637]
[ 0  1  2 88 99]
[ 437  210 1302 4641  485  991 2019 1263  312  171]


Now predict 10 movies for userId 1 using the top 50 most similar users

In [190]:
# Recommend 10 movies for userId 1
user_id = 1

user_ratings_mean = get_user_rating_mean(user_id)
top_50_most_similar_user_ids, similarity_scores = get_top_n_most_similar_users(1, 50)
similar_users_ratings = np.nan_to_num(get_ratings_for_users(top_50_most_similar_user_ids))
similar_users_ratings_mean = get_mean_ratings_for_users(top_50_most_similar_user_ids)
candidate_movie_ids = get_candidate_movie_ids(user_id, top_50_most_similar_user_ids)
candidate_movie_indices = get_movie_indices_for_movie_ids(candidate_movie_ids)

predicted_movie_indices, ratings = predict_movies(user_ratings_mean, similar_users_ratings, similar_users_ratings_mean, similarity_scores, candidate_movie_indices, 10)

print('predicted movie indices:')
print(predicted_movie_indices)
print()
print('predicted ratings:')
print(ratings)
print()

top_10_ratings_df = (
    pl.DataFrame({"index": predicted_movie_indices,
                  "predicted_rating": ratings})
)

# print top 10 movies
top_10_movies_df = (
    movies_df
    .filter(pl.col("index").is_in(predicted_movie_indices))
    .join(top_10_ratings_df, on="index")
    .drop("index")
)

print(top_10_movies_df)

predicted movie indices:
[ 277 1938 7355  123   31  507 2144  302  126  506]

predicted ratings:
[2.3414884 2.2917266 1.8985755 1.4971073 1.4671545 1.2913637 1.2901106
 1.2001786 1.1979115 1.1787481]

shape: (10, 4)
┌─────────┬─────────────────────────────────┬─────────────────────────────────┬──────────────────┐
│ movieId ┆ title                           ┆ genres                          ┆ predicted_rating │
│ ---     ┆ ---                             ┆ ---                             ┆ ---              │
│ i32     ┆ str                             ┆ str                             ┆ f32              │
╞═════════╪═════════════════════════════════╪═════════════════════════════════╪══════════════════╡
│ 318     ┆ Shawshank Redemption, The (199… ┆ Crime|Drama                     ┆ 2.341488         │
│ 2570    ┆ Walk on the Moon, A (1999)      ┆ Drama|Romance                   ┆ 2.291727         │
│ 78499   ┆ Toy Story 3 (2010)              ┆ Adventure|Animation|Children|C… ┆ 1.898576   

From the results we can see that the Pearson correlation function did not work quite well on itself. The best predicted rating is only 2.34 which is considered a bad rating.

#### d)

New similarity function

We chose the cosine similarity as a distance measure for the following reasons:
    - It is a common similarity measure for collaborative filtering
    - It is a good distance measure for sparse data

However, since the cosine similarity does not depend on magnitude of the vector it means that with our rating scale 1-5 the ratings of three movies [1, 1, 1] and [5, 5, 5] would be treated as eaqual in the cosine similarity calculation. This is of course false so we cannot use the plain cosine similarity measure for our application. What we do is to split the ratings into two groups, positive and negative ratings. We then calculate the cosine similarity for each group and combine them with a weighted average. We give more weight to the positive ratings since they are more important for the recommendation system.

We apply the same method of masking b to only include ratings that are rated by a in order to have a fair comparison.


In [167]:
from scipy.spatial import distance

def get_cosine_similarity(a, b):

    a_positive = np.copy(a)
    b_positive = np.copy(b)

    # mask out all ratings of user b which are not rated by user a
    mask = a_positive == 0
    b_positive[mask] = 0

    a_negative = np.copy(a_positive)
    b_negative = np.copy(b_positive)

    pos_threshold = 2.5
    # positive ratings cosine similarity
    a_positive[a_positive <= pos_threshold] = 0
    b_positive[b_positive <= pos_threshold] = 0

    positive_similarity = 0
    if np.sum(a_positive) != 0 and np.sum(b_positive) != 0:
        positive_similarity = 1 - distance.cosine(a_positive, b_positive)

    # negative ratings cosine similarity
    a_negative[a_negative > pos_threshold] = 0
    b_negative[b_negative > pos_threshold] = 0

    negative_similarity = 0
    if np.sum(a_negative) != 0 and np.sum(b_negative) != 0:
        negative_similarity = 1 - distance.cosine(a_negative, b_negative)

    # combine positive and negative similarity. Give more weight to positive similarity
    w_pos = 1.5
    w_neg = 1
    return (w_pos * positive_similarity + w_neg * negative_similarity) / (w_pos + w_neg)

# Test
a = np.array([4, 4, 0, 4, ])
b = np.array([5, 0, 3, 5])
c = np.array([1, 1, 1, 0])
d = np.array([3, 2, 3, 2])

cosine_similarity = lambda a, b: get_cosine_similarity(a, b)

print(cosine_similarity(a, b))
print(cosine_similarity(a, c))
print(cosine_similarity(a, d))
print(cosine_similarity(c, d))

0.4898979485566356
0.0
0.34641016151377546
0.23094010767585033


#### e)
User-based collaborative filtering for group recommendations.

In [168]:
# Create a group of users by picking the first 5 userIds
user_ids = [1, 2, 3, 4, 5]
# group = ratings_df.filter(pl.col("userId").is_in(userIds))
# print('Group:')
# print(group)
# print()


# Combine all the above functions to predict movies for a single user
def predict_movies_for_user(user_id, num_of_movies):
    similar_user_ids, similarity_scores = get_top_n_most_similar_users(user_id, 50)
    print(similarity_scores)

    user_ratings_mean = get_user_rating_mean(user_id)
    similar_users_ratings = np.nan_to_num(get_ratings_for_users(similar_user_ids))
    similar_users_ratings_mean = get_mean_ratings_for_users(similar_user_ids)
    candidate_movie_ids = get_candidate_movie_ids(user_id, similar_user_ids)
    candidate_movie_indices = get_movie_indices_for_movie_ids(candidate_movie_ids)

    return predict_movies(user_ratings_mean, similar_users_ratings, similar_users_ratings_mean, similarity_scores, candidate_movie_indices, num_of_movies)


n_movies = 500
predicted_movie_indices = np.ndarray(shape=(len(user_ids), n_movies), dtype=np.int32)
predicted_ratings = np.ndarray(shape=(len(user_ids), n_movies), dtype=np.float32)

for index, user_id in enumerate(user_ids):
    user_movie_indices, ratings = predict_movies_for_user(user_id, n_movies)
    predicted_movie_indices[index] = user_movie_indices
    predicted_ratings[index] = ratings

# Print top 10 movies for each user
print('Predicted movie indices:')
print(predicted_movie_indices[:, :10])
print()
print('Predicted ratings:')
print(predicted_ratings[:, :10])
print()


# Combine the predictions into a dataframe
predictions_df = (
    pl.DataFrame({"movie_index": predicted_movie_indices.reshape(-1),
                  "predicted_rating": predicted_ratings.reshape(-1),
                  "userId": np.array([[userId] * n_movies for userId in userIds]).reshape(-1)})
    .unique()
    )

# Find movies which have rating for all users in the group
movie_ratings_count = (
    predictions_df
    .group_by("movie_index")
    .agg(
        pl.count("predicted_rating")
        .alias("ratings_count")
    )
    .filter(pl.col("ratings_count") >= 5)
    .sort("ratings_count", descending=True)
)

# filter predictions_df to only include movies that have been rated by all users in the group
predictions_df = (
    predictions_df
    .filter(
        pl.col("movie_index")
        .is_in(movie_ratings_count.select("movie_index")
               )
    )
    .sort("movie_index")
    .pivot("movie_index", index="userId")
)
print(predictions_df)

[1.         1.         1.         1.         1.         0.9622505
 0.9258201  0.91855866 0.8783101  0.8728716  0.8669214  0.86640024
 0.8660254  0.8660254  0.8215838  0.8029551  0.80178374 0.7905694
 0.78693587 0.7745967  0.7676495  0.7592566  0.75592893 0.75
 0.7083333  0.7062815  0.7016464  0.6844476  0.65465367 0.6506001
 0.6437963  0.64122343 0.62469506 0.6172134  0.61237246 0.61237246
 0.61209774 0.6109598  0.6061281  0.6030227  0.6        0.5964321
 0.5962848  0.59487444 0.58760905 0.58389246 0.58345    0.56906945
 0.5673498  0.5669467 ]
[1.         1.         1.         1.         1.         1.
 1.         1.         1.         1.         1.         1.
 1.         0.9449112  0.8660254  0.8660254  0.80757284 0.7894737
 0.70636564 0.69337523 0.6875     0.6621222  0.6232876  0.5833333
 0.5833333  0.57735026 0.572822   0.56642324 0.54395807 0.53033006
 0.50536954 0.5037349  0.5        0.5        0.49408695 0.49407724
 0.46513027 0.4397966  0.4128614  0.39718696 0.3859225  0.378681
 

IndexError: index 9731 is out of bounds for axis 1 with size 9724

Average aggregated recommendations

In [ ]:
# Calculate the average rating for each movie
average_ratings_df = (
    predictions_df
    .drop("userId")
    .mean()
)

print("Average ratings:")
print(average_ratings_df)
print()

# find the top 10 movies based on the average rating
top_10_movies_avg_df = (
    average_ratings_df
    .transpose(include_header=True, header_name="index", column_names=["avg_rating"])
    .with_columns(
        pl.col("index")
        .cast(pl.Int32)
    )
    .sort("avg_rating", descending=True)
    .limit(10)
    .join(movies_df, on="index", how="left")
)

print("Top 10 movies:")
print(top_10_movies_avg_df)
print()

Least misery aggregated recommendations

In [ ]:
# Calculate least misery rating for each movie
least_misery_ratings_df = (
    predictions_df
    .drop("userId")
    .min()
)

print("Least misery ratings:")
print(least_misery_ratings_df)
print()

# find the top 10 movies based on the least misery rating
top_10_movies_least_misery_df = (
    least_misery_ratings_df
    .transpose(include_header=True, header_name="index", column_names=["least_misery_rating"])
    .with_columns(
        pl.col("index")
        .cast(pl.Int32)
    )
    .sort("least_misery_rating", descending=True)
    .limit(10)
    .join(movies_df, on="index", how="left")
)

print("Top 10 movies:")
print(top_10_movies_least_misery_df)
print()

In [ ]:
# Find movies that appear in both top 10 lists
common_movies = (
    top_10_movies_avg_df
    .join(top_10_movies_least_misery_df, on="index", how="inner")
    .select(["index", "title", "avg_rating", "least_misery_rating"])
)

print("Common movies:")
print(common_movies)

# Find movies that appear in only one of the top 10 lists
unique_movies_avg = (
    top_10_movies_avg_df
    .join(top_10_movies_least_misery_df, on="index", how="anti")
    .select(["index", "title", "avg_rating"])
)

print("Unique movies avg:")
print(unique_movies_avg)
print()

unique_movies_least_misery = (
    top_10_movies_least_misery_df
    .join(top_10_movies_avg_df, on="index", how="anti")
    .select(["index", "title", "least_misery_rating"])
)

print("Unique movies least misery:")
print(unique_movies_least_misery)
print()

From the previous results we can see that the top 10 movies are mostly same for the average and least misery aggregation methods. Both methods have 8 movies in common and 2 unique movies. The order of the common movies varies alot. This result is expected since the two methods are based on different principles.

From the predicted scores in both methods we can see that the predictions are not very good. This however, is not a fault of the measure but the group of users that we have chosen. The group of users do not seem to share a preference when it comes to movies. If we had chosen a group of users that have more similar taste, both methods would show better results.

The results we have now are both a tradeoff and will now leave at least one user disliking the movie the group would watch.

#### f)

Disagreement between the users in a group